### PR TITLE
feat: odk form component

### LINKED ISSUE #1924: [FEATURE] ODK integration
**What?**
Ability for the app to 
- interpret and display Excel ODK forms
- mechanism to save the collected data to the server

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [x] PR title descriptive (can be used in release notes)

## Description

Proof-of-concept `odk_form` template component to render odk forms.

**TODO**
- [x] Frontend ODK Component
- [x] Enketo form renderer integration
- [x] Example templates
- [x] Tools to transform xlsform locally to required enketo json
- [ ] Support image attachments
- [ ] Support form value preload/edit
- [ ] Form validation
- [ ] Documentation
- [ ] Data saving strategy (e.g. autosave all changes vs user click save form button)
- [ ] Review dependencies (node-forge not included, possibly some others required for widgets e.g. leaflet-draw)
- [ ] Review translation options
- [ ] XLSForm auto conversion (e.g. when syncing assets, would require way to identify and new api endpoint)

**Known Issues**
- Component adds requirement for a lot of extra 3rd party dependencies (e.g. jquery, bootstrap-datepicker). These could be reduced by removing/replacing specific widgets (e.g. datepicker), however this would require additional work and careful documentation in case full compatibility is not kept with ODK.

- Style files in particular are very large, requiring change to build budgets

- Leaflet-draw package not included as pinned to outdated version that is hard to include (so likely some aspects of map widget will not work - untested)

- ODK form files need to be converted before use. A standalone package has been added to do this, but requires developer to manually populate and use docker to convert (before populating to assets folder). See additional issues for this package in the created readme

- Image attachments not yet rendering

- Multi-page forms are all shown on a single screen 

## Review Notes
- Checkout branch and `yarn install`
- Create a form using preferred tools (e.g. excel, kobocollect builder). Recommend fairly basic forms only as advanced cases untested
- Copy form .xlsx file to `packages\@idemsInternational\xlsform-converter\forms` folder
- Run conversion script `yarn workspace @idemsInternational/xlsform-converter start`
- Copy generated output json file to app asset folder
- Adapt included `comp_odk` template to point to asset (will likely not work if changing path dynamically in component - to fix in future)
- Should see form rendered and response xml populated below

## Dev Notes
Originally it had been planned to build this as a standalone webcomponent and import, however as the standalone version is not yet ready for publishing the codebase has been copied and adapted to work directly in angular. Code in the `odk-form/libs` component folder has been copied directly from external repo and main component.ts adapted from it also

The benefit of this is further adjustments that could be made to customise the component and reduce bundle sizes. The drawback is loss of encapsulation for 3rd party libs (e.g. jquery) and harder long term maintainability

In terms of bundle size changes, it appears to be around 2MB increase overall (800kb node-forge, 300kb jquery, 900kb component js + few other smaller deps). Could consider replacing jquery with something like https://zeptojs.com/

**Local Changes**
In addition to copying over the main codebase, the following changes have been made:
- Update main component .ts to work with angular bindings

## Git Issues

Closes #1924

## Screenshots/Videos

[App.webm](https://github.com/IDEMSInternational/parenting-app-ui/assets/10515065/d218bcff-94ad-46de-855b-0cbb3847821c)

### CODE REVIEW

**Comment by volloholic** [comment]
@chrismclarke I have been discussing with Esmee and am not clear how https://xlsform.org/en/#multiple-choice-from-file would work? Does the population happen as part of the enketo Json creation process or is it still a live link then?

**Review by esmeetewinkel — CHANGES_REQUESTED** [review]
Workflow works fine for me, I've been able to set up docker / convert my own ODK `.xlsx` form / show it in the app (deployment [plh-facilitator](https://github.com/IDEMSInternational/plh-facilitator-app-content), template [odk_form_en](https://docs.google.com/spreadsheets/d/1HfQVzBWW4-LoclVGIhumk8aMjNOHJ-fueHetcxd2QXI/edit#gid=1321716080))

image

#### Comments: 
- The display looks quite funky. It seems that line breaks are not working, nor does conditionality. Also the text is black rather than inheriting the dark blue colour from our app theming. 
- How would I go about saving the outcome of a form to something we can access on the server, e.g. a contact field? 
- We would eventually need this to work with translation (which you observed already). When loading a multi-language ODK form into this at the moment, all languages are displayed (deployment [plh-facilitator](https://github.com/IDEMSInternational/plh-facilitator-app-content), template [odk_form](https://docs.google.com/spreadsheets/d/1HfQVzBWW4-LoclVGIhumk8aMjNOHJ-fueHetcxd2QXI/edit#gid=378849870))
image

**Comment by chrismclarke** [comment]
> @chrismclarke I have been discussing with Esmee and am not clear how https://xlsform.org/en/#multiple-choice-from-file would work? Does the population happen as part of the enketo Json creation process or is it still a live link then?

@volloholic 
Good question - I think we will need a minimal demo repo to see what the current behaviour is and then a discussion for what is required. As far as I can remember the csv is processed during form conversion, so only really supports a static list. I remember at Stats4SD we had to come up with a bunch of custom methods to recreate forms whenever supporting csvs changed to be able to populate with 'live' data.

Enketo does have support for custom widgets however, so if longer term we do need some extra functionality we could look into it

**Comment by chrismclarke** [comment]
@esmeetewinkel 
> The display looks quite funky. It seems that line breaks are not working, nor does conditionality. Select questions appear in side a (double) box, but no other question type does. Also the text is black rather than inheriting the dark blue colour from our app theming.

Just pushed a PR to fix the issue where enketo styling was not correctly being applied (new screenshot below). These enketo styles are quite complex and opinionated, so would recommend it being a follow-up task if we wanted to create our own custom theme to better inherit from the app styles.
image

> How would I go about saving the outcome of a form to something we can access on the server, e.g. a contact field?

The odk_form component is structures the same as any other input type component, so you can use the `changed` action trigger to save the value to a contact field. I can see from your demo you are already reading this value when populating the text at the bottom of the screen
image

The main issue is that enketo keeps all data written in xml format which isn't as easily interpreted as json would be. It's possible to convert xml to json, however I haven't done that up till now as it's quite hard to later reverse the process if we wanted to read back in the json record. E.g. the xml logged and stored in the value could be converted using the online converter
https://codebeautify.org/xmltojson
image

A short term workaround could be to store the result as both xml and json when we save (would require adding an extra 3rd party lib to convert like https://github.com/NaturalIntelligence/fast-xml-parser)

> We would eventually need this to work with translation (which you observed already). When loading a multi-language ODK form into this at the moment, all languages are displayed (deployment [plh-facilitator](https://github.com/IDEMSInternational/plh-facilitator-app-content), template [odk_form](https://docs.google.com/spreadsheets/d/1HfQVzBWW4-LoclVGIhumk8aMjNOHJ-fueHetcxd2QXI/edit#gid=378849870))

This should now be fixed with the style fixes
image

**Review by esmeetewinkel — APPROVED** [review]
Thanks @chrismclarke. Display looking much better, conditionality and translations work. I'm happy to leave the styling as-is for now. Approving this since I'm happy for this PR to be merged and to continue working on the issues flagged below in a follow-up PR. I'll leave it up to you to decide.

I'm able to save the XML output of the form to a field using the `changed` trigger. This saves the field every time any input of the form is changed. However, when using `start` and `end` in the ODK form the timestamps are both the same and do not change as I continue filling the form. (e.g. in Facilitator app deployment [ePLH_Pilot_Facilitator_Weekly_Survey_Combined.xlsx](https://github.com/IDEMSInternational/parenting-app-ui/files/11799058/ePLH_Pilot_Facilitator_Weekly_Survey_Combined.xlsx) = [converted json](https://drive.google.com/file/d/18lQPcLiC40Il6vXAysPJCMoOW10-QiN-/view?usp=drive_link))

Are we thinking of using the "Save Draft" and "Submit" buttons? If yes, how would I author their actions? If no, should we hide them (and use buttons in our regular syntax to achieve the same)?

image

**Comment by chrismclarke** [comment]
I've added #1926 to track issues flagged here.
So I think once @jfmcquade has reviewed should be good to go

**Review by jfmcquade — APPROVED** [review]
Looking good. Some minor comments/questions inline.

A few more comments:
- On narrow screens, the "Save Draft"/"Submit" buttons don't wrap or shrink so stick out of their container. I know there's a TODO to handle them better, so if they end up being kept then this styling should be fixed
Screenshot 2023-06-22 at 15 38 24

- Bundle size. You've included notes on what to remove (in `src/app/shared/components/template/components/odk-form/odk-form.component.ts`) if we want to "opt-out" of the ODK form feature, e.g. to reduce bundle size. I'm highlighting here mainly to make @esmeetewinkel aware that there is a brute force way to reduce the bundle size (by manually removing modules of code) if necessary for the next release of plh_tz, for example.

- It could be good to update the PR description to remove the TODO checklist, and link to #1964 instead

**Comment by esmeetewinkel** [comment]
Merging this as I need it for a deployment. Hopefully outstanding minor issues can be addressed in a follow-up PR.

File: .idems_app/deployments/plh_facilitator
[the remark this answers is not in the document]
Comment (chrismclarke) [reply .idems_app/deployments/plh_facilitator:1]: I think the files possibly were once there but then removed... I'll leave tidying it for a future PR

[the remark this answers is not in the document]
Comment (chrismclarke) [reply .idems_app/deployments/plh_facilitator:1]: resolved in #1926

File: src/app/shared/components/template/components/odk-form/odk-form.component.scss [hunk 1]
```diff
@@ -0,0 +1,73 @@
+@import "./libs/enketo/sass/grid/grid.scss";
+
+// apply :host style bindings for non-encapsulated component
+tmpl-odk-form {
+  background: white;
+  padding: 8px;
+  display: block;
+  position: relative;
+}
+// apply margin to fit section dropdown arrows (breakpoints from _responsiveness.scss)
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/components/odk-form/odk-form.component.scss:10]: resolved in #1926

File: src/app/shared/components/template/components/odk-form/odk-form.component.scss [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  margin-left: 30px;
+  @media screen and (max-width: 600px) {
+    margin-left: 24px;
+  }
+}
+#form-title {
+  margin: 0;
+}
+#form-unavailable {
+  // give form a chance to load/render before showing unavailable message
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/components/odk-form/odk-form.component.scss:21]: I was so sure that I must have copied this directly from the enketo-express implementation somewhere but can't seem to find it anymore so maybe that was all just a dream... In any case I think should serve our purpose fine for now

File: src/app/shared/components/template/components/odk-form/odk-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  private async handleEventDataUpdate(e: ReturnType<typeof Events.DataUpdate>) {
+    await this.setValue(this.xmlFormValue);
+    await this.triggerActions("changed");
+  }
+
+  private async handleEventXmlFormChange(e: ReturnType<typeof Events.XFormsValueChanged>) {
+    // Value saved in autoSave method of https://github.com/enketo/enketo-express/blob/master/public/js/src/module/controller-webform.js
+    // await this.setValue(this.xmlFormValue);
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/components/odk-form/odk-form.component.ts:115]: Yeah it's a bit confusing/messy to try and stay comparable with enketo, however I think I realised that the previous `handleEventDataUpdate` also triggers at the same time which currently handles our saveValue and triggerActions code, so possibly redundant. But may be reviewed when we also want to review the save draft/submit functionality

File: src/app/shared/components/template/components/odk-form/odk-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /**
+   * Load the form xml and data models and render intial form components
+   * https://enketo.github.io/enketo-core/tutorial-00-getting-started.html
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/components/odk-form/odk-form.component.ts:120]: Looks like it's moved, actually follows the code snippets from https://github.com/enketo/enketo-core#usage-as-a-library
resolved in #1926

File: packages/data-models/workflows/sync.workflows.ts [hunk 5]
```diff
@@ -81,14 +83,25 @@ const workflows: IDeploymentWorkflows = {
...
-        function: async ({ tasks, config }) =>
-          tasks.gdrive.download({
-            folderId: config.google_drive.assets_folder_id,
-            filterFn: config.google_drive.assets_filter_function,
-          }),
+        function: async ({ tasks, config, options }) => {
+          const folderId = config.google_drive.assets_folder_id;
+          const filterFn = config.google_drive.assets_filter_function;
+          // If skipping download still need to return download folder for next step
+          if (options.skipDownload) {
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply packages/data-models/workflows/sync.workflows.ts:99]: If I remember correclty all options defined at the start of the workflow are parsed using the same parser that [Commander.JS](https://www.npmjs.com/package/commander) uses, which by default turns options defined in `kebab-case` to `camelCase`. I assume this is for ease of authoring/convention, and is also the system we use for all the main workflow execution so makes sense to author in a similar way

File: src/app/shared/components/template/components/odk-form/odk-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ *
+ * Opt-Out (additional changes to make if removing component)
+ * - Remove modules:
+ *   jquery jquery-touchswipe mergexml/mergexml bootstrap-datepicker signature_pad html5sortable node-forge
+ * - Remove js file import in tsconfig.app.json
+ * - Reduce build budget styles 200kb -> 16kb
+ * - Reduce build budget js 5MB -> 4MB
+ */
+export class TmplOdkFormComponent extends TemplateBaseComponent implements OnInit {
+  private parameter_list: IODKFormComponentParameters;
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply src/app/shared/components/template/components/odk-form/odk-form.component.ts:69]: Resolved in #1926

**Pull request merged**